src/NumberTheory/QuadFormAndIsom/torsion_quadratic_module_with_isometry.jl

```diff
@@ -0,0 +1,229 @@
+#
+###############################################################################
+
+function torsion_quadratic_module_with_isometry(
+  T::TorQuadModule,
+  f::TorQuadModuleMap;
+  check::Bool=true,
+)
+  if check
+    @req domain(f) === codomain(f) === T "Wrong domain or codomain"
+    @req Hecke.is_isometry(f) "Not an isometry"
+  end
+
+  if isone(order(T))
+    return TorQuadModuleWithIsom(T, id_hom(T))
+  end
+
+  return TorQuadModuleWithIsom(T, f)
+end
+
+function torsion_quadratic_module_with_isometry(
+  T::TorQuadModule,
+  f::AutomorphismGroupElem{TorQuadModule};
+)
+  @req domain(f) === T "Wrong domain"
+  return torsion_quadratic_module_with_isometry(T, hom(f); check=false)
+end
+
+function torsion_quadratic_module_with_isometry(
+  T::TorQuadModule,
+  m::ZZMatrix;
+  check::Bool=true,
+)
+  f = hom(T, T, m; check)
+  return torsion_quadratic_module_with_isometry(T, f; check)
+end
+
+function torsion_quadratic_module_with_isometry(
+  T::TorQuadModule,
+  g::FinGenAbGroupHom;
+  check::Bool=true,
+)
+  if check
+    @req domain(g) === codomain(g) === abelian_group(T) "Wrong domain or codomain"
+  end
+  f = TorQuadModuleMap(T, T, g)
+  return torsion_quadratic_module_with_isometry(T, f; check)
+end
+
+function torsion_quadratic_module_with_isometry(
+  T::TorQuadModule,
+  g::MatrixGroupElem{QQFieldElem, QQMatrix};
+  check::Bool=true,
+)
+  if check
+    B = basis_matrix(relations(T))
+    @req can_solve(B, B*matrix(g); side=:left) "Incompatible inputs"
+  end
+  f = hom(T, T, elem_type(T)[T(lift(t)*matrix(g)) for t in gens(T)])
+  return torsion_quadratic_module_with_isometry(T, f; check)
+end
+
+function torsion_quadratic_module_with_isometry(
+  q::QQMatrix,
+  f::Union{TorQuadModuleMap, ZZMatrix, FinGenAbGroupHom, MatrixGroupElem{QQFieldElem, QQMatrix}};
+  check::Bool=true,
+)
+  T = TorQuadModule(q)
+  return torsion_quadratic_module_with_isometry(T, f; check)
+end
+
+###############################################################################
+#
+#  Submodules
+#
+###############################################################################
+
+function _sub(Tf::TorQuadModuleWithIsom, i::TorQuadModuleMap)
+  T = underlying_module(Tf)
+  f = isometry(Tf)
+  @req codomain(i) === T "Wrong codomain"
+  S = domain(i)
+  L = cover(S)
+  imgs = elem_type(S)[]
+  for a in gens(S)
+    b = lift(f(i(a)))
+    @req b in L "Submodule not preserved by the isometry"
+    push!(imgs, S(b))
+  end
+  g = hom(S, S, gens(S), imgs)
+  return torsion_quadratic_module_with_isometry(S, g; check=false)
+end
+
+function _is_submodule(Tf::TorQuadModuleWithIsom, S::TorQuadModule)
+  is_sublattice(cover(underlying_module(Tf)), cover(S)) || return false
+  f = isometry(Tf)
+  for a in gens(S)
+    lift(a) in cover(S) || return false
+  end
+  return true
+end
+
+function sub(
+  Tf::TorQuadModuleWithIsom,
+  gene::Vector{TorQuadModuleElem}
+)
+  T = underlying_module(Tf)
+  S, i = sub(T, gene)
+  Sg = _sub(Tf, i)
+  return Sg, i
+end
```

Comment:
You could add a more efficient version of this which is called when `check=false`?